### DISCUSSION TITLE
Failed to create instance

### DISCUSSION THREAD

**Discussion by Sirneij** [souffle-lang/souffle#2488] [Q&A] [answered]
I have this directory:

```sh
.
├── facts
│   └── edge.facts
├── test.cpp
├── test_souffle.dl
```

The contents of which are:

`facts/edge.facts`
```txt
1	2
1	3
1	4
1	5
1	6
1	7
1	8
1	9
```

`test.cpp`
```cpp
#include <souffle/SouffleInterface.h>
#include <iostream>
#include <string>

int main(int argc, char *argv[])
{
    using namespace std::chrono;

    if (argc < 3)
    {
        std::cerr << "Usage: " << argv[0] << " <datalog_file> <facts_file>" << std::endl;
        return 1;
    }

    const std::string datalogFile = argv[1];
    std::cout << "Datalog file: " << datalogFile << std::endl;
    const std::string factsFile = argv[2];
    std::cout << "Facts file: " << factsFile << std::endl;

    // Initialize Souffle program
    souffle::SouffleProgram *prog = souffle::ProgramFactory::newInstance(datalogFile);
    if (prog == nullptr)
    {
        std::cerr << "Could not create Souffle program instance." << std::endl;
        return 1;
    }

    // Load facts from file
    prog->loadAll(factsFile);

    // Run the program
    prog->run();

    // Retrieve relation "path" which contains the results of transitive closure
    souffle::Relation *path = prog->getRelation("path");
    if (path == nullptr)
    {
        std::cerr << "Relation 'path' not found." << std::endl;
        return 1;
    }

    // Iterate over the tuples in the relation
    for (auto &tuple : *path)
    {
        int start, end;
        tuple >> start >> end;
        std::cout << "Path: " << start << " -> " << end << std::endl;
    }

    // Cleanup
    delete prog;
    return 0;
}
```

`test_souffle.dl`
```prolog
.decl edge(x: number, y: number)
.input edge

.decl path(x: number, y: number)
.output path
path(x, y) :- edge(x, y).
path(x, y) :- path(x, z), edge(z, y).
```

When I issue the following command in my terminal:

```sh
souffle-compile.py test.cpp -o souffle_tmp
```

`souffle-compile.py` was installed alongside the souffle installation. I had used it to compile `.cpp` files generated by `souffle` before now without issues.

It compiles and `souffle_tmp` gets generated. However, when I want to run it:

```sh
./souffle_tmp ./test_souffle.dl ./facts/edge.facts
```

or

```sh
./souffle_tmp test_souffle.dl facts/edge.facts
```

I get this error:

```sh
Datalog file: test_souffle.dl
Facts file: facts/edge.facts


Could not create Souffle program instance.
```

What I can think of is that the souffle instance could not be created. Is something wrong with my code or something?

**Comment by quentin** [comment, marked as answer]
Hi,
`newInstance` does not take a datalog file name, it takes the name of a soufflé program (look at the generated c++ to get the exact identifier).

**Reply by Sirneij** [reply]
Thanks for the quick response. How then do we load the Datalog and fact files? And if I generate C++ code from the files, `newInstance` has nothing passed in:

```cpp
namespace  souffle {
using namespace souffle;
class factory_Sf_tmp: souffle::ProgramFactory {
public:
souffle::SouffleProgram* newInstance();
 factory_Sf_tmp();
private:
};
} // namespace  souffle
namespace  souffle {
using namespace souffle;
souffle::SouffleProgram* factory_Sf_tmp::newInstance(){
return new  souffle::Sf_tmp();
}

 factory_Sf_tmp::factory_Sf_tmp():
souffle::ProgramFactory("tmp"){
}

} // namespace  souffle
```

Also, does it mean that one can't write a normal C++ program and use souffle to compile it without first generating a C++ code out of the Datalog programs? What does the C++ interface (https://souffle-lang.github.io/interface) do?

**Reply by quentin** [reply]
That's exact, the C++ interface provides functions to interact with c++ code generated by Souffle: populating input relations, running the program, reading output relations.

**Reply by Sirneij** [reply]
Thank you so much for the enlightment. I am sorry for the initial knowledge gap. I generated C++ from the datalog programs and used the interface to interact with the generated file. The name passed to the `newInstance` was the generated file's name, without its extension, and that worked.
